sdks/python/apache_beam/runners/interactive/extensions/apache-beam-jupyterlab-sidepanel/src/common/HtmlView.tsx

```diff
@@ -0,0 +1,119 @@
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+import * as React from 'react';
+
+export interface IHtmlProvider {
+  readonly html: string;
+  readonly script: Array<string>;
+}
+
+interface IHtmlViewProps {
+  htmlProvider: IHtmlProvider;
+}
+
+interface IHtmlViewState {
+  innerHtml: string;
+  script: Array<string>;
+}
+
+/**
+ * A common HTML viewing component that renders given HTML and executes scripts
+ * from the given provider.
+ */
+export class HtmlView extends React.Component<IHtmlViewProps, IHtmlViewState> {
+  constructor(props: IHtmlViewProps) {
+    super(props);
+    this.state = {
+      innerHtml: props.htmlProvider.html,
+      script: []
+    };
+  }
+
+  componentDidMount(): void {
+    this._updateRenderTimerId = setInterval(() => this.updateRender(), 1000);
+  }
+
+  componentWillUnmount(): void {
+    clearInterval(this._updateRenderTimerId);
+  }
+
+  updateRender(): void {
+    const currentHtml = this.state.innerHtml;
+    const htmlToUpdate = this.props.htmlProvider.html;
+    const currentScript = this.state.script;
+    const scriptToUpdate = [...this.props.htmlProvider.script];
+    if (htmlToUpdate !== currentHtml) {
+      this.setState({
+        innerHtml: htmlToUpdate,
+        // As long as the html is updated, clear the script state.
+        script: []
+      });
+    }
+    /* Depending on whether this iteration updates the html, the scripts
+     * are executed differently.
+     * Html updated: all scripts are new, start execution from index 0;
+     * Html not updated: only newly added scripts need to be executed.
+     */
+    const currentScriptLength =
+      htmlToUpdate === currentHtml ? currentScript.length : 0;
+    if (scriptToUpdate.length > currentScriptLength) {
+      this.setState(
+        {
+          script: scriptToUpdate
+        },
+        // Executes scripts once the state is updated.
+        () => {
+          for (let i = currentScriptLength; i < scriptToUpdate.length; ++i) {
+            new Function(scriptToUpdate[i])();
+          }
+        }
+      );
+    }
+  }
+
+  render(): React.ReactNode {
+    return (
+      // This injects raw HTML fetched from kernel into JSX.
+      <div dangerouslySetInnerHTML={{ __html: this.state.innerHtml }} />
+    );
+  }
+
+  private _updateRenderTimerId: number;
+}
+
+/**
+ * Makes the browser support HTML import and import HTML from given hrefs if
+ * any is given.
+ *
+ * Native HTML import has been deprecated by modern browsers. To support
+ * importing reusable HTML templates, webcomponentsjs library is needed.
+ * The given hrefs will be imported once the library is loaded.
+ *
+ * Note everything is appended to head and if there are duplicated HTML
+ * imports, only the first one will take effect.
+ */
+export function importHtml(hrefs: Array<string>): void {
```

Comment:
Is this only included for test purposes? Or is there somewhere else that it's expected to be used.

Reply:
This is needed when initializing the `Widget` during activation of the extension.

The `importHtml` only needs to be invoked once.
The HTML we are targeting is the one used by [PAIR-facets](https://pair-code.github.io/facets/).
Once invoked, later in the view, those facets visualization sent back from the kernel could be rendered correctly.